File: ios-app/src/tasks/taskChainNdef.ts
import { Ndef, NdefRecord } from 'react-native-nfc-manager';
import { utf8Decode, utf8Encode } from '../nfc/textCodec';
import { TaskChain } from './types';

/** A TaskChain can ride along on a tag as an extra MIME record, so scanning it both shows
 * the tag's normal NDEF content and offers to run the attached automation. */
export const TASKS_MIME_TYPE = 'application/vnd.nfckit.tasks+json';

export function taskChainToRecord(chain: TaskChain): NdefRecord {
  return Ndef.record(Ndef.TNF_MIME_MEDIA, TASKS_MIME_TYPE, [], utf8Encode(JSON.stringify(chain)));
}

export function findTaskChain(records: NdefRecord[]): TaskChain | null {
  const typeToString = (type: number[] | string) => (typeof type === 'string' ? type : Ndef.util.bytesToString(type));
  const record = records.find((r) => r.tnf === Ndef.TNF_MIME_MEDIA && typeToString(r.type) === TASKS_MIME_TYPE);
  if (!record) return null;
  try {
    return JSON.parse(utf8Decode(Uint8Array.from(record.payload)));
  } catch {
    return null;
  }
}

File: app/src/main/java/com/zephyrcloud/nfckit/tasks/TaskChainNdef.kt
package com.zephyrcloud.nfckit.tasks

import android.nfc.NdefMessage
import android.nfc.NdefRecord
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * A [TaskChain] can ride along on a tag as an extra MIME record so that scanning the tag
 * with NFC Kit both shows its normal NDEF content *and* offers to run the attached
 * automation -- this is the "tag triggers a task" flow from the Tasks tab.
 */
object TaskChainNdef {
    const val MIME_TYPE = "application/vnd.nfckit.tasks+json"

    private val json = Json { ignoreUnknownKeys = true }

    fun toRecord(chain: TaskChain): NdefRecord =
        NdefRecord(
            NdefRecord.TNF_MIME_MEDIA,
            MIME_TYPE.toByteArray(Charsets.US_ASCII),
            ByteArray(0),
            json.encodeToString(chain).toByteArray(Charsets.UTF_8),
        )

    fun findInMessage(message: NdefMessage): TaskChain? {
        val record = message.records.firstOrNull {
            it.tnf == NdefRecord.TNF_MIME_MEDIA && String(it.type, Charsets.US_ASCII) == MIME_TYPE
        } ?: return null
        return try {
            json.decodeFromString<TaskChain>(String(record.payload, Charsets.UTF_8))
        } catch (_: Exception) {
            null
        }
    }
}
